encoding/gzip/gzip.go

```diff
@@ -27,49 +27,73 @@ import (
 	"google.golang.org/grpc/encoding"
 )
 
-// New function registers gzip compressor. It maintains two pools for compressor and decompressor.
-func New() encoding.Compressor {
-	return &compressor{
+func init() {
+	encoding.RegisterCompressor(&compressor{
 		name: "gzip_register",
 		poolCompressor: sync.Pool{
 			New: func() interface{} {
-				return gzip.NewWriter(ioutil.Discard)
+				return &gzipWriter{w: gzip.NewWriter(ioutil.Discard)}
 			},
 		},
 		poolDecompressor: sync.Pool{},
-	}
+	})
+}
+
+type gzipWriter struct {
+	w    *gzip.Writer
+	pool *sync.Pool
 }
 
 func (c *compressor) Compress(w io.Writer) (io.WriteCloser, error) {
-	z := c.poolCompressor.Get().(*gzip.Writer)
-	z.Reset(w)
+	z := c.poolCompressor.Get().(*gzipWriter)
+	z.pool = &c.poolCompressor
+	z.w.Reset(w)
 	return z, nil
 }
 
+func (z *gzipWriter) Close() error {
+	defer z.pool.Put(z)
+	return z.w.Close()
+}
+
+func (z *gzipWriter) Write(p []byte) (int, error) {
+	return z.w.Write(p)
+}
+
+type gzipReader struct {
+	r    *gzip.Reader
+	pool *sync.Pool
+}
+
 func (c *compressor) Decompress(r io.Reader) (io.Reader, error) {
-	var z *gzip.Reader
+	var z *gzipReader
 	switch maybeZ := c.poolDecompressor.Get().(type) {
 	case nil:
 		newZ, err := gzip.NewReader(r)
 		if err != nil {
 			return nil, err
 		}
-		z = newZ
-	case *gzip.Reader:
+		z = &gzipReader{r: newZ, pool: &c.poolDecompressor}
+	case *gzipReader:
 		z = maybeZ
-		if err := z.Reset(r); err != nil {
+		if err := z.r.Reset(r); err != nil {
 			c.poolDecompressor.Put(z)
 			return nil, err
 		}
 	}
-	defer func() {
-		c.poolDecompressor.Put(z)
-	}()
 	return z, nil
```

Comment:
With both paths above returning, this is no longer possible.  A panic should be fine here instead of a return.

Alternatively, it may be possible to restructure so the nil case happens outside the switch (or use an if).